crates/fuel-core/src/service/adapters/fuel_gas_price_provider.rs

```diff
@@ -0,0 +1,225 @@
+pub struct FuelGasPriceProvider<FB, DA, A, GP> {
+    profitablility_totals: ProfitablilityTotals,
+
+    block_history: FB,
+    da_recording_cost_history: DA,
+    algorithm: A,
+    gas_price_history: GP,
+}
+
+impl<FB, DA, A, GP> FuelGasPriceProvider<FB, DA, A, GP> {
+    pub fn new(
+        block_history: FB,
+        da_recording_cost_history: DA,
+        algorithm: A,
+        gas_price_history: GP,
+    ) -> Self {
+        Self {
+            profitablility_totals: ProfitablilityTotals::default(),
+            block_history,
+            da_recording_cost_history,
+            algorithm,
+            gas_price_history,
+        }
+    }
+}
+
+struct ProfitablilityTotals {
+    totaled_block_height: RefCell<BlockHeight>,
+    total_reward: RefCell<u64>,
+    total_cost: RefCell<u64>,
+}
+
+impl Default for ProfitablilityTotals {
+    fn default() -> Self {
+        Self {
+            totaled_block_height: RefCell::new(0.into()),
+            total_reward: RefCell::new(0),
+            total_cost: RefCell::new(0),
+        }
+    }
+}
+
+impl ProfitablilityTotals {
+    #[allow(dead_code)]
+    pub fn new(block_height: BlockHeight, reward: u64, cost: u64) -> Self {
+        Self {
+            totaled_block_height: RefCell::new(block_height),
+            total_reward: RefCell::new(reward),
+            total_cost: RefCell::new(cost),
+        }
+    }
+
+    fn update(&self, block_height: BlockHeight, reward: u64, cost: u64) {
+        let mut totaled_block_height = self.totaled_block_height.borrow_mut();
+        let mut total_reward = self.total_reward.borrow_mut();
+        let mut total_cost = self.total_cost.borrow_mut();
+        *totaled_block_height = block_height;
+        *total_reward += reward;
+        *total_cost += cost;
+    }
+
+    fn totaled_block_height(&self) -> BlockHeight {
+        *self.totaled_block_height.borrow()
+    }
+}
+
+impl<FB, DA, A, GP> FuelGasPriceProvider<FB, DA, A, GP>
+where
+    FB: FuelBlockHistory,
+    DA: DARecordingCostHistory,
+    A: GasPriceAlgorithm,
+    GP: GasPriceHistory,
+{
+    fn inner_gas_price(&self, requested_block_height: BlockHeight) -> Result<u64> {
+        let latest_block = self
+            .block_history
+            .latest_height()
+            .map_err(Error::UnableToGetLatestBlockHeight)?;
+        if latest_block > requested_block_height {
+            let gas_prices = self
+                .gas_price_history
+                .gas_prices(requested_block_height)
+                .map_err(Error::UnableToGetGasPrices)?
+                .ok_or(Error::GasPricesNotFoundForBlockHeight(
+                    requested_block_height,
+                ))?;
+            Ok(gas_prices.total())
+        } else if Self::asking_for_next_block(latest_block, requested_block_height) {
+            let new_gas_prices = self.calculate_new_gas_price(latest_block)?;
+            self.gas_price_history
+                .store_gas_prices(requested_block_height, &new_gas_prices)
```

Comment:
This was `store_gas_prices(latest_block)` before and the tests were still passing. Fixing it didn't break the tests either, so the coverage needs to be updated if we want to keep this approach.